routes.py

```diff
@@ -0,0 +1,217 @@
+
+
+@routes.route(api_routes_base + '/<provider>', methods=["GET"])
+def link_info(provider):
+    user_info = auth.require_user_info(request)
+    refresh_token = _get_provider(provider).bond.get_link_info(user_info)
+    if refresh_token:
+        return protojson.encode_message(LinkInfoResponse(issued_at=refresh_token.issued_at, username=refresh_token.username))
+    else:
+        raise exceptions.NotFound("{} link does not exist".format(provider))
+
+
+@routes.route(api_routes_base + '/<provider>', methods=["DELETE"])
+def delete_link(provider):
+    user_info = auth.require_user_info(request)
+    _get_provider(provider).bond.unlink_account(user_info)
+    return protojson.encode_message(message_types.VoidMessage()), 204
+
+
+@routes.route(api_routes_base + '/<provider>/accesstoken', methods=["GET"])
+def accesstoken(provider):
+    user_info = auth.require_user_info(request)
+    try:
+        access_token, expires_at = _get_provider(provider).bond.generate_access_token(user_info)
+        return protojson.encode_message(AccessTokenResponse(token=access_token, expires_at=expires_at))
+    except Bond.MissingTokenError as err:
+        # TODO: I don't like throwing and rethrowing exceptions
+        raise exceptions.BadRequest(err.message)
+
+
+@routes.route(api_routes_base + '/<provider>/serviceaccount/key', methods=["GET"])
+def service_account_key(provider):
+    user_info = auth.require_user_info(request)
+    return protojson.encode_message(ServiceAccountKeyResponse(data=json.loads(
+        _get_provider(provider).fence_tvm.get_service_account_key_json(user_info))))
+
+
+@routes.route(api_routes_base + '/<provider>/serviceaccount/accesstoken', methods=["GET"])
+@use_args({"scopes": fields.Str(repeated=True, missing=None)},
+          locations=("querystring",))
+def service_account_accesstoken(args, provider):
+    user_info = auth.require_user_info(request)
+    return protojson.encode_message(ServiceAccountAccessTokenResponse(token=_get_provider(provider).fence_tvm.get_service_account_access_token(user_info, args['scopes'])))
+
+
+@routes.route(api_routes_base + '/<provider>/authorization-url', methods=["GET"])
+@use_args({"scopes": fields.Str(repeated=True, missing=None),
+           "redirect_uri": fields.Str(required=True),
+           "state": fields.Str(missing=None)},
+          locations=("querystring",))
+def authorization_url(args, provider):
+    authz_url = _get_provider(provider).bond.build_authz_url(args['scopes'], args['redirect_uri'], args['state'])
+    return protojson.encode_message((AuthorizationUrlResponse(url=authz_url)))
+
+
+@routes.route('/api/status/v1/status', methods=["GET"])
+def get_status():
+    sam_base_url = config.get('sam', 'BASE_URL')
+
+    providers = {provider_name: FenceApi(config.get(provider_name, 'FENCE_BASE_URL'))
+                 for provider_name in config.sections() if provider_name != 'sam'}
```

Comment:
this looks pretty similar to:

```
bond_providers = {provider_name: create_provider(provider_name)
                  for provider_name in config.sections() if provider_name != 'sam'}
```

but I don't have an idea of how to optimize that and may be out of scope for this ticket

Reply:
Is that code from somewhere else? I'm not really sure what we'd be optimizing here, and I believe that this stuff was transferred over 1-1 from the old `main.py`